content/blog/2023-12-11-rvv-llvm-gisel/index.md

```diff
@@ -0,0 +1,177 @@
+
+  Michael Maitland was a student in Cornell CS 6120 Advanced Compilers and he's now working at SiFive.
+"""
+[[extra.authors]]
+name = "Jiahan Xie"
+[[extra.authors]]
+name = "Michael Maitland"
++++
+
+# Introduction
+
+The open [RISC-V instruction set architecture (ISA)](https://riscv.org/technical/specifications/) has an interesting extension, [the RISC-V "V" Vector Extension](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc). The unique part of RISC-V Vector Extension is that its vector registers have flexible widths, VLEN, which makes programming in RISC-V Vector Extension agnostic to the vector register sizes. This feature really distinguishes the RISC-V Vector Extension from the traditional SIMD extensions, such as [x86 SSE](https://en.wikipedia.org/wiki/Streaming_SIMD_Extensions) (with a fixed size 128-bit vector length)/[AVX](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions) (256-bit), and [Arm NEON](https://developer.arm.com/Architectures/Neon) (128-bit). The increasing vector lengths can pose a challenge to the traditional SIMD extensions as they have to address compatability and support all existing fixed size vector lengths in their ISAs. On the contrary, with the vector lengths agnostic principles, binary code generated by RISC-V assembly is automatically portable between different CPUs.
+
+# What Was the Goal?
+The goal was to support LLVM Global Instruction Selection for RISC-V Vector Extensions on some ALU operations, such as `vadd`, `vsub`, `vand`, `vor`, and `vxor`.
+
+
+# A Primer on the RISC-V Vector Extension
+There are a lot of instructions and cool features in the [RISC-V "V" Vector Extension Spec](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc), I will only pick the most relevant and important parts to explain. 
+
+The most important parameters are undoubtly `SEW`, `VLEN`, `VL`, and `LMUL`; and one of the most interesting and powerful instructions is `vset{i}vl{i}`.
+
+Let's begin with the crucial parameters:
+- `SEW`: Selected Element Width (in bits), set dynamically by the programmers. It sets the width/length of a single element in a vector element. Each vector element can compose multiple single element.
+- `VLEN`: Vector register LENgth (in bits). The number of bits in a single vector register. It is hardware dependent. 
+- `VL`: Vector element Length (in bits) that the programmers actually deal with, which can be treated as the vector operation building blocks. It defines how many elements the vector operations will execute.
+- `LMUL`: The vector Length MULtiplier. It is used for grouping vector registers. It is a power of 2 and it ranges from 1/8 to 8. For instance, when `LMUL=8`, only `v0`, `v8`, `v16`, and `v24` indices are allowed to used, as for example, group `v8` encodes 8 vector elments `v8v9`...`v15`. Note it can also be fraction numbers because sometimes we want to use only parts of the vector registers.
```

Comment:
Technically the ABI imposes the fact that `v0`, `v8`, `v16` can be used, not the RISC-V Vector spec. You could have used `v1, v9, v17`. The reason the ABI does it the way it does is so that v17-v31 can be used in the caller/callee save scheme. The important part here is that on LMUL 8, the register `vN` acts as a grouping of `vN, vN+1, ..., vN+7` registers. That forces us to partition what registers can be passed and preserve the expected semantics of the instruction.

Reply:
That is correct. According to the spec:

```
When LMUL=2, the vector register group contains vector register v n and vector register v n+1, providing twice the vector length in bits. Instructions specifying an LMUL=2 vector register group with an odd-numbered vector register are reserved.

When LMUL=4, the vector register group contains four vector registers, and instructions specifying an LMUL=4 vector register group using vector register numbers that are not multiples of four are reserved.

When LMUL=8, the vector register group contains eight vector registers, and instructions specifying an LMUL=8 vector register group using register numbers that are not multiples of eight are reserved.
```

Reply:
Got it; thanks!